Gentoo does something different than upstream. This patch allows multiple
directories to be sourced. Upstream is trying to come to a consensus on this
issue and it is expected to change. Therefore it is a moot point to send this
upstream now.

--- bash_completion.sh.orig	2009-04-08 23:14:34.736632335 -0500
+++ bash_completion.sh	2009-04-08 23:17:59.670875954 -0500
@@ -4,9 +4,46 @@
 # Check for recent enough version of bash.
 bash=${BASH_VERSION%.*}; bmajor=${bash%.*}; bminor=${bash#*.}
 if [ $bmajor -eq 2 -a $bminor '>' 04 ] || [ $bmajor -gt 2 ]; then
-  if [ -r /etc/bash_completion ]; then
-    # Source completion code.
-    . /etc/bash_completion
-  fi
+    _load_completions() {
+    declare f x loaded_pre=false
+    for f; do
+        if [[ -f $f ]]; then
+        # Prevent loading base twice, initially and via glob
+        if $loaded_pre && [[ $f == */base ]]; then
+            continue
+        fi
+
+        # Some modules, including base, depend on the definitions
+        # in .pre.  See the ebuild for how this is created.
+        if ! $loaded_pre; then
+            if [[ ${BASH_COMPLETION-unset} == unset ]]; then
+            BASH_COMPLETION=/usr/share/bash-completion/base
+            fi
+            source /usr/share/bash-completion/.pre
+            loaded_pre=true
+        fi
+
+        source "$f"
+        fi
+    done
+
+    # Clean up
+    $loaded_pre && source /usr/share/bash-completion/.post
+    unset -f _load_completions  # not designed to be called more than once
+    }
+
+    # 1. Load base, if eselected.  This was previously known as
+    #    /etc/bash_completion
+    # 2. Load completion modules, maintained via eselect bashcomp --global
+    # 3. Load user completion modules, maintained via eselect bashcomp
+    # 4. Load user completion file last, overrides modules at user discretion
+	# This order is subject to change once upstream decides on something.
+    _load_completions \
+    /etc/bash_completion.d/base \
+    ~/.bash_completion.d/base \
+    /etc/bash_completion.d/* \
+    ~/.bash_completion.d/* \
+    ~/.bash_completion
 fi
+
 unset bash bminor bmajor
